### PR TITLE
Refactor Trainer and Model

### PR DESCRIPTION
**Pull request by ervteng**
Refactor trainer, model in preparation for SAC. 

Specifically:
- Move common functions to `trainer.py`, model.py` from `ppo/trainer.py`, `ppo/policy.py` and `ppo/model.py'
- Introduce `RLTrainer` class and move most of `add_experiences` and some common reward 
signal code there. PPO and SAC will inherit from this, not so much BC Trainer.
- Add methods to Buffer to enable sampling, truncating, and save/loading. 
- Add scoping to create encoders in `model.py`

### CODE REVIEW

File: ml-agents/mlagents/trainers/buffer.py [hunk 1]
```diff
@@ -240,6 +282,22 @@ def reset_update_buffer(self):
         """
         self.update_buffer.reset_agent()
 
+    def truncate_update_buffer(self, max_length):
+        """
+        Truncates the update buffer to a certain length.
+
+        The skew makes a bigger truncation each time, hopefully requiring less truncations overall.
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/buffer.py]: What is the skew?

File: ml-agents/mlagents/trainers/trainer.py [hunk 2]
```diff
@@ -268,3 +281,237 @@ def write_tensorboard_text(self, key, input_dict):
                 "Cannot write text summary for Tensorboard. Tensorflow version must be r1.2 or above."
             )
             pass
+
+
+class RLTrainer(Trainer):
```
Comment (awjuliani) [inline ml-agents/mlagents/trainers/trainer.py]: Does it make sense to make this its own file?

File: ml-agents/mlagents/trainers/buffer.py (hunk 1 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/buffer.py]: Seems like an old comment from an older implementation - removed.

File: ml-agents/mlagents/trainers/trainer.py (hunk 2 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/trainer.py]: Was debating this - I think it does, but I was also concerned with creating more files. Seems like it's long enough to break off, though.

Comment (ervteng) [reply ml-agents/mlagents/trainers/trainer.py]: Done

File: ml-agents/mlagents/trainers/models.py [hunk 3]
```diff
@@ -407,7 +409,7 @@ def create_discrete_action_masking_layer(all_logits, action_masks, action_size):
             for i in range(len(action_size))
         ]
         raw_probs = [
-            tf.multiply(tf.nn.softmax(branches_logits[k]) + 1.0e-10, branch_masks[k])
+            tf.multiply(tf.nn.softmax(branches_logits[k]) + 1.0e-7, branch_masks[k])
```
Comment (harperj) [inline ml-agents/mlagents/trainers/models.py]: Can you explain the change from `1.0e-10` to `1.0e-7` here and below?

File: ml-agents/mlagents/trainers/models.py [hunk 4]
```diff
@@ -438,12 +438,15 @@ def create_observation_streams(
         h_size: int,
         num_layers: int,
         vis_encode_type: EncoderType = EncoderType.SIMPLE,
+        stream_scopes: List[str] = [],
```
Comment (harperj) [inline ml-agents/mlagents/trainers/models.py]: We shouldn't use default values of `[]` -- if two separate models are created using the defaults and one changes the value of `stream_scopes`, it will change for the other.

File: ml-agents/mlagents/trainers/models.py (hunk 3 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/models.py]: In SAC I found that 1.0e-10 was too small to prevent NaNs. I'm changing it here to be safe, though I don't think there were any known NaNs for PPO.

File: ml-agents/mlagents/trainers/models.py (hunk 4 above)
Comment (ervteng) [reply ml-agents/mlagents/trainers/models.py]: Fixed, changed to None

**Comment by ervteng** [comment]
Fixed a couple bugs (mostly to do with SAC) I've tested with 3DBall, Hallway, and GridWorld.
